app/src/main/java/com/github/fribourgsdp/radio/database/FirestoreDatabase.kt

```diff
@@ -1,563 +1,645 @@
-            val snapshot = transaction.get(docRef)
-
-            if (!snapshot.exists()) {
-                throw IllegalArgumentException("Document $gameID in games not found.")
-            }
-
-            // reset done map
-         
-            val updatedDoneMap = snapshot.getPlayerDoneMap()
-            updatedDoneMap.replaceAll { _, _ -> false}
-
-            // reset found map
-            val updatedFoundMap = snapshot.getPlayerFoundMap()
-            updatedFoundMap.replaceAll { _, _ -> false}
-
-            // reset scores of round
-            val scoresOfRound = snapshot.getScoresOfRound<Long>()
-            scoresOfRound.replaceAll { _, _ -> 0L}
-
-            // Update on database
-            transaction.update(
-                docRef,
-                "player_done_map", updatedDoneMap,
-                "player_found_map", updatedFoundMap,
-                "scores_of_round", scoresOfRound
-            )
-
-            // Success
-            null
-        }
-    }
-
-    private fun listenUpdate(collectionPath : String, id: Long, listener: EventListener<DocumentSnapshot>){
-        db.collection(collectionPath).document(id.toString())
-            .addSnapshotListener(listener)
-    }
+package com.github.fribourgsdp.radio.database
+
+import android.content.ContentValues
+import android.util.Log
+import com.github.fribourgsdp.radio.*
+import com.github.fribourgsdp.radio.data.*
+import com.github.fribourgsdp.radio.game.Game
+import com.github.fribourgsdp.radio.util.*
+import com.google.android.gms.tasks.Task
+import com.google.android.gms.tasks.Tasks
+import com.google.firebase.Timestamp
+import com.google.firebase.firestore.DocumentReference
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.ktx.firestore
+import com.google.firebase.ktx.Firebase
+import com.google.firebase.firestore.EventListener
+import com.google.firebase.firestore.FieldValue
+import java.lang.Exception
+import java.util.*
+import kotlin.collections.ArrayList
+import kotlin.collections.HashMap
+
+
+/**
+ * This class serves to make possible the dependency injection with mockito. We mock the
+ * reference and serve documents snapshot that are also mocked.
+ *
+ */
+open class FirestoreRef {
+    val db = Firebase.firestore
+    open fun getUserRef(userId : String) : DocumentReference{
+        return db.collection("user").document(userId)
+    }
+    open fun getSongRef(songId : String) : DocumentReference{
+        return db.collection("songs").document(songId)
+    }
+    open fun getPlaylistRef(playlistId : String) : DocumentReference{
+        return db.collection("playlists").document(playlistId)
+    }
+}
+/**
+ *
+ *This class represents a database. It communicates with the database(Firestore)
+ * and translates the result in classes
+ *
+ * @constructor Creates a database linked to Firestore
+ */
+class FirestoreDatabase(var refMake: FirestoreRef) : Database {

```

Comment:
For some reason, Github thinks I rewrote the entirety of this file. This is one of the reasons why it says I have written that many lines of code and it also impacted the diff coverage.